telluric/products.py

```diff
@@ -0,0 +1,840 @@
+    return [item for sublist in main_list for item in sublist]
+
+
+Product = namedtuple('Product', ['raster', 'bands_mapping', 'name', 'display_name', 'description',
+                                 'min', 'max', 'type', 'required_bands',
+                                 'output_bands', 'unit', 'default_view', 'used_bands'])
+
+
+class _Product():
+    """ this class is the result of applying a product on a raster"""
+
+    def __init__(self, raster, bands_mapping, product_generator, **overrides):
+        product_generator.update(overrides)
+        self.raster = raster
+        self.bands_mapping = bands_mapping
+        self.name = product_generator['name']
+        self.display_name = product_generator['display_name']
+        self.description = product_generator['description']
+        self.min = product_generator['min']
+        self.max = product_generator['max']
+        self.type = product_generator['type']
+        self.required_bands = product_generator['required_bands']
+        self.output_bands = product_generator['output_bands']
+        self.unit = product_generator['unit']
+        self.default_view = product_generator['default_view']
+
+    def used_bands(self):
+        return flatten_two_level_list(self.bands_mapping.values())  # flatten bands list
+
+    def namedtuple(self):
+        p = Product(
+            raster=self.raster,
+            bands_mapping=self.bands_mapping,
+            name=self.name,
+            display_name=self.display_name,
+            description=self.description,
+            min=self.min,
+            max=self.max,
+            type=self.type,
+            required_bands=self.required_bands,
+            output_bands=self.output_bands,
+            unit=self.unit,
+            default_view=self.default_view,
+            used_bands=self.used_bands()
+        )
+        return p
+
+
+class ProductError(Exception):
+    pass
+
+
+class ProductsFactory(BaseFactory):
+    __objects = None
+
+    @classmethod
+    def objects(cls):
+        if not cls.__objects:
+            subclasses = sorted(ProductGenerator.__subclasses__(), key=lambda p: (p.get_order(), p.get_name().lower()))
+            cls.__objects = OrderedDict()
+            for p in subclasses:
+                if p.dont_add_to_factory:
+                    continue
+                cls.__objects[p.get_name().lower()] = p
+        return cls.__objects
+
+
+class ProductGenerator:
+    """ProductGenerator is an abstract class.
+
+    To implement a product class you should inherit from it.
+    You should implement _apply
+    You should set the following attributes:
+        name min max required_bands output_bands description default_view type display_name unit _order
+    """
+
+    should_convert_to_float = True
+
+    # don't add this Product generator to ProductsFactory list
+    dont_add_to_factory = False
+
+    # dicts with the wavelengths for products band definitions
+    wavelength_map = {
```

Comment:
I don't think we should expose our band names here,

Reply:
this are not our band names except, red, green, blue, nir, and pan which are universal names

Reply:
R405 is a global name?

Reply:
yes, it is not a name that we use in our settings, it is a general name for that band as it is refer in the definition of the indices

Reply:
And why did you choose these definitions?